File: backend/enum-jpa-validation-app/src/main/java/com/example/enumapp/web/validation/NotEmptyList.java
package com.example.enumapp.web.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import jakarta.validation.ReportAsSingleViolation;
import jakarta.validation.constraints.NotEmpty;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/** GET {@code @RequestParam List} 용 — null/empty 금지. */
@Documented
@Target({ElementType.FIELD, ElementType.PARAMETER, ElementType.TYPE_USE})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = {})
@NotEmpty
@ReportAsSingleViolation
public @interface NotEmptyList {

    String message() default "list must not be null or empty";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: backend/enum-jpa-querydsl-kotlin-app/src/main/kotlin/com/example/enumapp/web/validation/NotEmptyList.kt
package com.example.enumapp.web.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import jakarta.validation.ReportAsSingleViolation
import jakarta.validation.constraints.NotEmpty
import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.FIELD
import kotlin.annotation.AnnotationTarget.TYPE
import kotlin.annotation.AnnotationTarget.VALUE_PARAMETER
import kotlin.reflect.KClass

/** GET `@RequestParam List` 용 — null/empty 금지. */
@MustBeDocumented
@Target(FIELD, VALUE_PARAMETER, TYPE)
@Retention(RUNTIME)
@Constraint(validatedBy = [])
@NotEmpty
@ReportAsSingleViolation
annotation class NotEmptyList(
    val message: String = "list must not be null or empty",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
)
